### PR TITLE
initial SYCL build system and examples

### PR DESCRIPTION
**Pull request by yxlao**
- Initial build system with SYCL
- Python unit test, C++ unit test, C++ example, C++ linkage with as third-party lib are all working
- CPU only, not tested with iGPU or dGPU
- See `3rdparty/README_SYCL.md` on how to set up the SYCL environment

This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/5081)

### CODE REVIEW

**Comment by yxlao** [comment]
Closing, in favor of #4902

File: 3rdparty/assimp/assimp.cmake [hunk 1]
```diff
@@ -6,6 +6,12 @@ else()
     set(lib_name assimp)
 endif()
 
+if(CMAKE_CXX_COMPILER_ID MATCHES "IntelLLVM")
+    set(assimp_cmake_cxx_flags ${CMAKE_CXX_FLAGS} -fno-fast-math)
```
Comment (ssheorey) [inline 3rdparty/assimp/assimp.cmake:10]: comment why?

File: 3rdparty/benchmark/MakeAvailable/CMakeLists.txt [hunk 2]
```diff
@@ -9,6 +9,8 @@ else()
     add_compile_definitions(_GLIBCXX_USE_CXX11_ABI=0)
 endif()
 
+add_compile_options($<$<CXX_COMPILER_ID:IntelLLVM>:-fno-fast-math>)
+
```
Comment (ssheorey) [inline 3rdparty/benchmark/MakeAvailable/CMakeLists.txt:13]: add comment for why?

File: CMakeLists.txt [hunk 3]
```diff
@@ -59,7 +64,12 @@ if(BUILD_SHARED_LIBS)
 else()
     option(STATIC_WINDOWS_RUNTIME "Use static (MT/MTd) Windows runtime"      ON )
 endif()
-option(GLIBCXX_USE_CXX11_ABI      "Set -D_GLIBCXX_USE_CXX11_ABI=1"           ON )
+option(BUILD_SYCL_MODULE          "Build SYCL module with Intel oneAPI"      OFF)
+if(BUILD_SYCL_MODULE)
+    option(GLIBCXX_USE_CXX11_ABI  "Set -D_GLIBCXX_USE_CXX11_ABI=1"           ON )
+else()
+    option(GLIBCXX_USE_CXX11_ABI  "Set -D_GLIBCXX_USE_CXX11_ABI=1"           OFF)
+endif()
```
Comment (ssheorey) [inline CMakeLists.txt:72]: make GLIBCXX_USE_CXX11_ABI dependent option for BUILD_SYCL_MODULE ?

File: CMakeLists.txt [hunk 4]
```diff
@@ -248,9 +258,32 @@ project(Open3D
...
+    message(FATAL_ERROR "BUILD_SYCL_MODULE=ON requires BUILD_PYTORCH_OPS=OFF")
+endif()
+
 # Global flag to set CXX standard.
 # This does not affect 3rd party libraries.
-set(CMAKE_CXX_STANDARD 14)
+if (BUILD_SYCL_MODULE)
+    set(CMAKE_CXX_STANDARD 17)
+else()
+    set(CMAKE_CXX_STANDARD 14)
```
Comment (ssheorey) [inline CMakeLists.txt:285]: Perhaps we should set this to 17 here, and set it to 14 only for the tf_ops folder? The C++ version shouldn't matter during linking.

File: .github/workflows/ubuntu-sycl.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  NPROC: 2
+  GCE_CLI_GHA_VERSION: "302.0.0"
+
+jobs:
+  ubuntu-sycl:
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+      matrix:
+        SHARED: [ON, OFF]
```
[the remark this answers is not in the document]
Comment (yxlao) [reply .github/workflows/ubuntu-sycl.yml:24]: The current solution has the advantage that the command is explicit, i.e. you can directly copy and paste the same command to run on a local system. 

```bash
docker/docker_build.sh sycl-shared
docker/docker_build.sh sycl-static
```

This is also the approach used in `ubuntu.yml`.

File: 3rdparty/assimp/assimp.cmake (hunk 1 above)
Comment (yxlao) [reply 3rdparty/assimp/assimp.cmake:10]: Added.

File: 3rdparty/benchmark/MakeAvailable/CMakeLists.txt (hunk 2 above)
Comment (yxlao) [reply 3rdparty/benchmark/MakeAvailable/CMakeLists.txt:13]: Added.

File: CMakeLists.txt (hunk 3 above)
Comment (yxlao) [reply CMakeLists.txt:72]: Dependent options are hard to read. This one is more verbose but clear.

File: CMakeLists.txt (hunk 4 above)
Comment (yxlao) [reply CMakeLists.txt:285]: Our previous effort to migrate to C++17 (https://github.com/isl-org/Open3D/pull/4698) hit some roadblocks due to TensorFlow. This shall be addressed further in a separate PR.

**Review by benjaminum — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/5081)* status: 0 of 37 files reviewed, 6 unresolved discussions (waiting on @benjaminum, @ssheorey, and @yxlao)

___
*[`3rdparty/find_dependencies.cmake` line 1377 at r6](https://reviewable.io/reviews/isl-org/open3d/5081#-N2VnNspA5GB_ZpsHbOs:-N2VnNspA5GB_ZpsHbOt:b-e8fqje) ([raw file](https://github.com/isl-org/open3d/blob/2b40539564012e9fa7260d2d86e580f43e5667b6/3rdparty/find_dependencies.cmake#L1377)):*
> ```CMake
>         message(STATUS "Using OneAPI packages from: $ENV{ONEAPI_ROOT}")
>     else()
>         message(FATAL_ERROR "OneAPI packages not found. Please run `source /opt/intel/oneapi/setvars.sh`")
> ```

Just a comment, the env var is not set for the conda distribution and is not used to find the packages as far as I know.
We can update when the conda problems with oneAPI are solved.

**Review by yxlao — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/5081)* status: 0 of 37 files reviewed, 6 unresolved discussions (waiting on @benjaminum and @ssheorey)

___
*[`3rdparty/find_dependencies.cmake` line 1377 at r6](https://reviewable.io/reviews/isl-org/open3d/5081#-N2VnNspA5GB_ZpsHbOs:-N2XEKrB-dXOiv6Ddoao:b2oo0wp) ([raw file](https://github.com/isl-org/open3d/blob/2b40539564012e9fa7260d2d86e580f43e5667b6/3rdparty/find_dependencies.cmake#L1377)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Just a comment, the env var is not set for the conda distribution and is not used to find the packages as far as I know.
We can update when the conda problems with oneAPI are solved.
</blockquote></details>

Sounds good, I'll remove this check for now.

**Pull request merged**